File: python/pr_2_2.py
import time


start_time = time.perf_counter()


cache = {0: 0, 1: 1}


def fibo(n):
    if n in cache.keys():
        return cache[n]
    else:
        cache[n] = fibo(n - 1) + fibo(n - 2)
    return cache[n]


i = 0
fibo_list = []
while fibo(i) < 4000_000:
    fibo_list.append(fibo(i))
    i += 1

even_fibo = filter(lambda a: a % 2 == 0, fibo_list)
print(sum(list(even_fibo)))
end_time = time.perf_counter()
elapsed_time = end_time - start_time
print(f"Elapsed time is {elapsed_time}")

File: cpp/pr_2_2.cpp
#include <chrono> // for std::chrono functions
#include <iostream>
#include <vector>

class Timer {
private:
  // Type aliases to make accessing nested type easier
  using Clock = std::chrono::steady_clock;
  using Second = std::chrono::duration<double, std::ratio<1>>;

  std::chrono::time_point<Clock> m_beg{Clock::now()};

public:
  void reset() { m_beg = Clock::now(); }

  double elapsed() const {
    return std::chrono::duration_cast<Second>(Clock::now() - m_beg).count();
  }
};

struct Cache {
  int key{};
  int value{};
};

int fibonacci(std::vector<Cache> &cache, int n) {
  if (n <
      static_cast<int>(cache.size())) // checking if this value of n is computed
    return cache.data()[n].value;
  else
    cache.push_back({n, fibonacci(cache, n - 1) +
                            fibonacci(cache, n - 2)}); // if not compute
  return cache[static_cast<std::size_t>(n)].value;
}

int evenSum(std::vector<Cache> &array) {
  int sum{};
  for (Cache element : array)
    if (element.value % 2 == 0)
      sum += element.value;
  return sum;
}

int main() {
  Timer t;
  int i{0};
  std::vector<Cache> cache{{0, 1}, {1, 1}};

  while (cache.back().value < 400000000) {
    fibonacci(cache, i);
    ++i;
  }
  cache.pop_back(); // gotta take out the last one (should have written btter
                    // but whatever)
  std::cout << evenSum(cache) << '\n';

  std::cout << "Time elapsed: " << t.elapsed() << " seconds\n";

  return 0;
}
